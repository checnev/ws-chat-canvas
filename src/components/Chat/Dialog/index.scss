.dialog {
  position: relative;
  flex-grow: 1;
  min-height: 100px;

  &__window {
    position: absolute;
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    width:100%;
    height:100%;
    top: 0;
    left: 0;
  }

  &__messages {
    overflow-y: auto;
    padding: 10px;
  }
}
