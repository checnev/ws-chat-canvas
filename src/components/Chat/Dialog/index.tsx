import React from 'react';
import cn from 'utils/cn';
import Message, { IMessage } from './Message';
import './index.scss';
import type User from 'lib/user/User';

interface DialogProps {
  messages: IMessage[];
  user: User;
}

class Dialog extends React.Component<DialogProps> {
  messagesRef = React.createRef<HTMLDivElement>();

  componentDidUpdate() {
    const messages = this.messagesRef.current;

    messages?.scrollBy({
      top: messages.scrollHeight,
      behavior: 'smooth',
    });
  }

  render() {
    const dialog = cn('dialog');

    return (
      <div className={dialog()}>
        <div className={dialog('window')}>
          <div className={dialog('messages')} ref={this.messagesRef}>
            {this.props.messages.map((message) => (
              <Message
                key={message.id}
                direction={message.author.id === this.props.user.id ? 'out' : 'in'}
                message={message}
              />
            ))}
          </div>
        </div>
      </div>
    );
  }
}

export default Dialog;
