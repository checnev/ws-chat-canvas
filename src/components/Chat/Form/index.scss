@import '~material-icons/iconfont/material-icons.scss';

.chatForm {
  position: relative;
  height: 60px;
  border-top: 1px solid rgba(0, 0, 0, .1);

  &__submit {
    @include material-icons();
    @include material-icon('send');
    position: absolute;
    display: block;
    width: 36px;
    height: 36px;
    right: 10px;
    top: 10px;
    border:none;
    font-size: 24px;
    color: rgb(30, 167, 253);
    background: rgb(255, 255, 255);
    cursor: pointer;
  }

  textarea {
    width: 100%;
    height: 100%;
    resize: none;
    padding: 8px 12px;
    border: none;
  }
}
