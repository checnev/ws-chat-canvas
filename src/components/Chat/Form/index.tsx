import React from 'react';
import cn from 'utils/cn';
import { Form, FormikProps } from 'formik';
import './index.scss';

export interface ChatFormValues {
  text: string;
}

type ChatFormProps = FormikProps<ChatFormValues>;

class ChatForm extends React.Component<ChatFormProps> {

  render() {
    const chatForm = cn('chatForm');
    return (
      <Form
        className={chatForm()}
      >
        <textarea
          name="text"
          placeholder="What's new ?"
          onChange={this.props.handleChange}
          onBlur={this.props.handleBlur}
          value={this.props.values.text}
        />
        <button
          type="submit"
          className={chatForm('submit')}
        />
      </Form>
    );
  }
}

export default ChatForm;
