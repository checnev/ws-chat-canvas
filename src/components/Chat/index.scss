.chat {
  height: 100%;
  border: 1px solid rgba(0, 0, 0, .1);

  .gridRow {
    height: 100%;

    .gridRow__column:first-child {
      display: flex;
      flex-direction: column;
      justify-content: flex-end;
    }
  }
}
