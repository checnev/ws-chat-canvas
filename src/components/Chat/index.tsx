import React from 'react';
import cn from 'utils/cn';
import { GridRow, GridColumn } from 'components/common/Grid';
import { Dialog, Form } from 'containers/Chat';
import './index.scss';


class Chat extends React.Component {
  render() {
    const chat = cn('chat');
    return (
      <div className={chat()}>
        <GridRow>
          <GridColumn xs={12}>
            <Dialog />
            <Form />
          </GridColumn>
        </GridRow>
      </div>
    );
  }
}

export default Chat;
