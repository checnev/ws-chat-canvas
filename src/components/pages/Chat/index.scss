.chatPage {
  height: 100%;
  padding-top: 30px;

  .container {
    height: 100%;
  }

  @media screen and (max-width: 600px) {
    &__online {
      margin-top: 20px;
    }
  }
}
