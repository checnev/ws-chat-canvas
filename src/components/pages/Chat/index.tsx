import React from 'react';
import cn from 'utils/cn';
import Container from 'components/common/Container';
import { GridRow, GridColumn } from 'components/common/Grid';
import Chat from 'components/Chat'
import OnlineUserList from 'containers/OnlineUserList';
import './index.scss';

class ChatPage extends React.Component {
  
  render() {
    const chatPage = cn('chatPage');
    return (
      <div className={chatPage()}>
        <Container maxWidth="xl">
          <GridRow>
            <GridColumn xs={12} sm={9}>
              <Chat />
            </GridColumn>
            <GridColumn xs={12} sm={3}>
              <div className={chatPage('online')}>
                <OnlineUserList />
              </div>
            </GridColumn>
          </GridRow>
        </Container>
      </div>
    );
  }
}

export default ChatPage;
