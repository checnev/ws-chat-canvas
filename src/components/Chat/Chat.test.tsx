import { render, unmountComponentAtNode } from 'react-dom';
import { act } from "react-dom/test-utils";
import pretty from 'pretty';
import Chat from 'components/Chat';

let container: any = null;

beforeEach(() => {
  container = document.createElement('div');
  document.body.append(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

it('should render Chat', () => {
  act(() => {
    render(<Chat />, container);
  });

  expect(
    pretty(container.innerHTML)
  ).toMatchSnapshot();
});
