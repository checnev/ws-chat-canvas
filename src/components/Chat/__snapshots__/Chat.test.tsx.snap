// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Chat 1`] = `
"<div class=\\"chat\\">
  <div class=\\"gridRow\\">
    <div class=\\"gridRow__column gridRow__column_xs_12\\">
      <div class=\\"dialog\\">
        <div class=\\"dialog__window\\">
          <div class=\\"dialog__messages\\"></div>
        </div>
      </div>
      <form action=\\"#\\" class=\\"chatForm\\"><textarea name=\\"text\\" placeholder=\\"What's new ?\\"></textarea><button type=\\"submit\\" class=\\"chatForm__submit\\"></button></form>
    </div>
  </div>
</div>"
`;
